File: house-robber.java
public class Solution {
    public int rob(int[] nums) {
        /*动态规划：
        dp[i] = max(dp[i-2] + nums[i], dp[i-1]);
        */
        if (nums == null || nums.length == 0) return 0;
        int[] dp = new int[nums.length];
        for (int i = 0; i < nums.length; ++i) {
            dp[i] = Integer.max((i-2>=0?dp[i-2]:0) + nums[i], (i-1)>=0?dp[i-1]:0);
        }
        return dp[nums.length - 1];
    }
}

File: src/main/cpp/house-robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        if (nums.size() == 0) {
            return 0;
        } else if (nums.size() <= 2) {
            return nums.size() == 1 ? nums[0] : max(nums[0], nums[1]);
        }
        vector<int> dp;
        dp.push_back(nums[0]);
        dp.push_back(max(nums[0], nums[1]));
        for (int i = 2; i < nums.size(); i++) {
            dp.push_back(max(dp[i-1], dp[i-2] + nums[i]));
        }
        return dp[nums.size() - 1];
    }
};